[Fix] `no-unused-modules`: don't error out when running with flat config and an eslintrc isn't present

michaelfaith opened the pull request:
This change adjusts what we're doing if an error is thrown while attempting to enumerate lintable files in the `no-used-modules` rule, when users are running with flat config.  Now, if `FileEnumerator` throws due to a lack of eslintrc _and_ the user is running with flat config, we catch the error and rethrow with a more informative message.

I also cleaned up the original aspects of the implementation that was using the proposed eslint context functions, since that proposal was walked back and the API was never introduced.

Note: This isn't an ideal state, since this rule still relies on the legacy api to understand what needs to be ignored.  Remaining tethered to the legacy config system is going to need to be solved at some point.

Fixes #3079

ljharb commented on .gitignore:
```diff
@@ -52,3 +52,6 @@ npm-shrinkwrap.json
 # macOS
 .DS_Store
+
+# vscode settings
+.vscode
```
this should probably go in a global gitignore rather than in every individual project

michaelfaith replied:
Some projects I work in do keep some workspace settings in source control.  This is the only one where I need to actively turn off formatting.  I can remove it, if you prefer.  it's just a bit of a pain to always have to remember to turn it off.

ljharb replied:
hmm - i'd expect those projects would explicitly *un*ignore it in their gitignore since so many people globally gitignore vscode settings already. certainly we can keep it in here if it's that important to you.

ljharb commented on src/rules/no-unused-modules.js:
```diff
@@ -51,21 +50,81 @@ function requireFileEnumerator() {
+
+  // This condition is sufficient to test in v8, since the environment variable
+  // is necessary to turn on flat config
+  let isUsingFlatConfig =
+    process.env.ESLINT_USE_FLAT_CONFIG &&
+    process.env.ESLINT_USE_FLAT_CONFIG !== 'false';
```
```suggestion
  let isUsingFlatConfig = process.env.ESLINT_USE_FLAT_CONFIG &&
    process.env.ESLINT_USE_FLAT_CONFIG !== 'false';
```
linebreaks after the `=` is a signature prettier thing :-(

michaelfaith replied:
Updated

ljharb reviewed (commented):
Looks pretty good, thanks!

michaelfaith commented:
Addressed lint errors

michaelfaith commented:
Updated, based on feedback in #3079 to not attempt to avoid the error, but to instead rethrow with a more informative message.  I updated the PR description to reflect the new approach.

ljharb commented:
I suppose we could have a test make a temp dir (using `tmp`), copy an example into it, run `npm install` in it, and then invoke eslint inside it?

michaelfaith commented:
> I suppose we could have a test make a temp dir (using `tmp`), copy an example into it, run `npm install` in it, and then invoke eslint inside it?

Added a test that does the following:
- copies the v9 example folder
- runs exec to install the plugin from cwd
- run the lint command
- catches the error and validates that the message includes our new verbiage

michaelfaith commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@ This change log adheres to standards from [Keep a CHANGELOG](https://keepachange
 - add TypeScript types ([#3097], thanks [@G-Rath])
 
+### Fixed
+- [`no-unused-modules`]: don't error out when running with flat config and an eslintrc isn't present ([#3116], thanks [@michaelfaith])
```
I think this changelog entry will need to be revised, but I didn't want to adjust that since you added it.

ljharb replied:
totally fine either way; the format should hopefully be self-explanatory based on the commit message

michaelfaith replied:
Updated

michaelfaith commented:
@ljharb I got this about where I want it, except for one question to you.  I added the test to the existing set of unit tests just because all the CI plumbing was already in place to add it there.  Though, it doesn't really feel right for it to be there, since it's more of an integration test.  Would you rather it be where I have it now, or add a new workflow to CI for doing this kind of integration testing?

ljharb commented:
I think that it's fine where it is to land this, but it might make sense to move it in a followup?

michaelfaith commented:
> I think that it's fine where it is to land this, but it might make sense to move it in a followup?

Sounds good

> … although i do see there's 2 uncovered lines in the diff

Do you mean the catch block?  If so, that's because the new test isn't really contributing to the coverage report, since it's building the plugin and running lint as a separate exec command.
